pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -18,38 +18,220 @@ package cm
 
 import (
 	"fmt"
+	"strings"
+	"sync"
 
 	"github.com/golang/glog"
 
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/sets"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
 	"k8s.io/kubernetes/pkg/kubelet/deviceplugin"
 )
 
+// podDevices represents a list of pod to Device mappings.
+type containerToDevice map[string]sets.String
+type podDevices struct {
+	podDeviceMapping map[string]containerToDevice
+}
+
+func newPodDevices() *podDevices {
+	return &podDevices{
+		podDeviceMapping: make(map[string]containerToDevice),
+	}
+}
+func (pdev *podDevices) pods() sets.String {
+	ret := sets.NewString()
+	for k := range pdev.podDeviceMapping {
+		ret.Insert(k)
+	}
+	return ret
+}
+
+func (pdev *podDevices) insert(podUID, contName string, device string) {
+	if _, exists := pdev.podDeviceMapping[podUID]; !exists {
+		pdev.podDeviceMapping[podUID] = make(containerToDevice)
+	}
+	if _, exists := pdev.podDeviceMapping[podUID][contName]; !exists {
+		pdev.podDeviceMapping[podUID][contName] = sets.NewString()
+	}
+	pdev.podDeviceMapping[podUID][contName].Insert(device)
+}
+
+func (pdev *podDevices) getDevices(podUID, contName string) sets.String {
+	containers, exists := pdev.podDeviceMapping[podUID]
+	if !exists {
+		return nil
+	}
+	devices, exists := containers[contName]
+	if !exists {
+		return nil
+	}
+	return devices
+}
+
+func (pdev *podDevices) delete(pods []string) {
+	for _, uid := range pods {
+		delete(pdev.podDeviceMapping, uid)
+	}
+}
+
+func (pdev *podDevices) devices() sets.String {
+	ret := sets.NewString()
+	for _, containerToDevice := range pdev.podDeviceMapping {
+		for _, deviceSet := range containerToDevice {
+			ret = ret.Union(deviceSet)
+		}
+	}
+	return ret
+}
+
+// updateCapacityCallback is called to update ContainerManager capacity when
+// device capacity changes.
+type updateCapacityCallback func(v1.ResourceList)
+
 type DevicePluginHandlerImpl struct {
-	devicePluginManager deviceplugin.Manager
+	sync.Mutex
+	devicePluginManager *deviceplugin.ManagerImpl
+	// allDevices and allocated are keyed by resource_name.
+	allDevices map[string]sets.String
+	allocated  map[string]*podDevices
+	updateCapacity updateCapacityCallback
 }
 
 // NewDevicePluginHandler create a DevicePluginHandler
-func NewDevicePluginHandler() (*DevicePluginHandlerImpl, error) {
+func NewDevicePluginHandler(enabled bool, updateCapacity updateCapacityCallback) (*DevicePluginHandlerImpl, error) {
+	if !enabled {
```

Comment:
I wonder if should use the same pattern as for the GPU manager. As in:
`type DevicePluginHandler interface`
`type DevicePluginHandlerImpl struct`
`type MockDevicePluginHandler struct`

And when calling NewDevicePluginHandler instead of passing a bool:
```golang
if enabled {
cm.handler = NewDevicePluginHandlerImpl()
} else {
cm.handler = NewMockDevicePluginHandler()
}